trie.c

```diff
@@ -110,59 +131,84 @@ bool st_find_longest_chain(st_cursor_t *cursor, st_trie_match_t *longest_match,
     const st_trie_t *trie = cursor->trie;
     bool longer_match_found = false;
     do {
-        uint8_t code = TDATA(trie, offset);
-        st_assert(code, "Unexpected null code! Offset: %d", offset);
-        st_assert(!(code & TRIE_MATCH_BIT), "Match found at top of loop! Offset: %d", offset);
+        st_assert(TDATA(trie, offset), "Unexpected null code! Offset: %d", offset);
+        st_trie_node_info_t node_info;
+        st_get_node_info(trie, &node_info, &offset);
 
-        // Branch Node (with multiple children) if bit 14 is set
-        if (code & TRIE_BRANCH_BIT) {
-            st_debug(ST_DBG_SEQ_MATCH, "Branching Offset: %d; Code: %#04X", offset, code);
-            code = TDATA(trie, ++offset);
+        // Match Node if bit 15 is set
+        if (node_info.has_match) {
+            if (node_info.has_unchained_match) {
+                st_debug(ST_DBG_SEQ_MATCH, "New Match found: (%d, %d) %d\n",
+                    cursor->pos.index, cursor->pos.sub_index, cursor->pos.segment_len);
+                st_debug(ST_DBG_SEQ_MATCH, "Previous Match: (%d, %d) %d\n",
+                    longest_match->seq_match_pos.index, longest_match->seq_match_pos.sub_index, longest_match->seq_match_pos.segment_len);
+                // record this if it is the longest match
+                if (st_cursor_longer_than(cursor, &longest_match->seq_match_pos)) {
+                    longer_match_found = true;
+                    longest_match->trie_match_index = offset;
+                    longest_match->seq_match_pos = st_cursor_save(cursor);
+                }
+                offset += TRIE_MATCH_SIZE;
+            }
+            if (node_info.chain_check_count > 0) {
+                uint16_t match_index = st_cursor_get_matched_rule(cursor);
+                st_debug(ST_DBG_SEQ_MATCH, "Checking for sub-rule matching %#06X\n", match_index);
+                if (match_index != ST_DEFAULT_KEY_ACTION) {
+                    for (int i = 0; i < node_info.chain_check_count; i++) {
+                        const uint16_t sub_rule_match_index = st_get_trie_data_word(trie, offset);
+                        st_debug(ST_DBG_SEQ_MATCH, "  sub-rule %#06X\n", sub_rule_match_index);
+                        if (match_index == sub_rule_match_index) {
+                            // This sub-rule was previously matched. This chained rule
+                            // must be the longest match, so we record it and return immediately
+                            // The match index is at offset + 2
+                            // (sub-rule-byte1 sub-rule-byte2 match-byte1 match-byte2 match-byte3 match-byte4)
+                            longest_match->trie_match_index = offset + 2;
+                            longest_match->seq_match_pos = st_cursor_save(cursor);
+                            longest_match->is_chained_match = true;
```

Comment:
A chained match is always considered final. This is done for two reasons. One, it is impossible for a any rule to truly be longer than a chained match. Second, we don't track how many characters a rule matched in the sequence, so we are unable to accurately produce a true seq_match_len value for chained rules, so we need a way to signal "trust me, this is the longest match possible, stop searching".

This does mean that the current implementation of `log rule` will produce garbage output for chained rules. But, that will be replaced with a much simpler function that will simply output the index of the rule match, and the companion app will translate that into a human readable rule.